File: chatbot/backend/src/seed.ts
/**
 * Seed Script — Pre-computes embeddings for all knowledge base chunks.
 *
 * Usage:
 *   1. Set OPENAI_API_KEY environment variable
 *   2. Run: npx tsx src/seed.ts
 *   3. Use the output JSON to update the knowledge chunks with embeddings
 *
 * This script should be run before deploying the worker, and the resulting
 * embeddings can be embedded directly into the worker code for faster cold starts.
 */

import { writeFileSync } from 'fs'
import { buildKnowledgeBase } from './rag/knowledge'
import { embedBatch } from './rag/embedder'

const OPENAI_API_KEY = process.env.OPENAI_API_KEY
if (!OPENAI_API_KEY) {
  console.error('Error: OPENAI_API_KEY environment variable is required')
  console.error('Usage: OPENAI_API_KEY=sk-... npx tsx src/seed.ts')
  process.exit(1)
}

const mockEnv = {
  OPENAI_API_KEY,
  EMBEDDING_MODEL: 'text-embedding-3-small',
} as any

async function main() {
  console.log('Building knowledge base...')
  const chunks = buildKnowledgeBase()
  console.log(`Generated ${chunks.length} knowledge chunks`)

  console.log('\nGenerating embeddings via OpenAI...')
  const texts = chunks.map((c) => c.content)
  const embeddings = await embedBatch(texts, mockEnv)
  console.log(`Generated ${embeddings.length} embeddings`)

  // Attach embeddings to chunks
  const enrichedChunks = chunks.map((chunk, i) => ({
    ...chunk,
    embedding: embeddings[i],
  }))

  // Output as JSON for embedding into the worker code
  const outputPath = './src/rag/embeddings.json'
  writeFileSync(outputPath, JSON.stringify(enrichedChunks, null, 2))

  console.log(`\n✅ Saved embeddings to ${outputPath}`)
  console.log('Embedding dimension:', embeddings[0].length)
  console.log('Total characters:', texts.reduce((a, b) => a + b.length, 0))
}

main().catch((err) => {
  console.error('Seed failed:', err)
  process.exit(1)
})

File: chatbot/backend-python/src/seed.py
"""
Seed Script — Pre-computes embeddings for all knowledge base chunks
and caches them to disk for faster startup.

Usage:
    cd chatbot/backend-python
    python src/seed.py

This script:
1. Builds the knowledge base
2. Computes embeddings using sentence-transformers (local, no API key needed)
3. Saves to src/rag/embeddings_cache.json

On subsequent runs, the chatbot loads the cached embeddings for instant startup.
"""

import json
import os
import sys
from pathlib import Path

# Ensure we can import from the project src directory
_project_root = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(_project_root))

from src.rag.knowledge import build_knowledge_base
from src.rag.embedder import embed_batch


CACHE_PATH = os.path.join(
    os.path.dirname(os.path.abspath(__file__)),
    "rag",
    "embeddings_cache.json",
)


def main():
    print("Building knowledge base...")
    chunks = build_knowledge_base()
    print(f"Generated {len(chunks)} knowledge chunks")

    print("\nComputing embeddings via sentence-transformers (all-MiniLM-L6-v2)...")
    print("(First run downloads the model ~80MB, subsequent runs use cache)")
    texts = [c.content for c in chunks]
    embeddings = embed_batch(texts, show_progress=True)
    print(f"Generated {len(embeddings)} embeddings (dim={len(embeddings[0])})")

    # Build serializable cache
    cache_data = [
        {
            "id": chunk.id,
            "content": chunk.content,
            "category": chunk.category,
            "metadata": chunk.metadata,
            "embedding": emb,
        }
        for chunk, emb in zip(chunks, embeddings)
    ]

    with open(CACHE_PATH, "w") as f:
        json.dump(cache_data, f, indent=2)

    print(f"\n✅ Saved embeddings cache to {CACHE_PATH}")
    print(f"   File size: {os.path.getsize(CACHE_PATH) / 1024:.1f} KB")
    print(f"   Embedding dimension: {len(embeddings[0])}")
    print(f"   Total characters: {sum(len(t) for t in texts)}")


if __name__ == "__main__":
    main()
